### DISCUSSION TITLE
Is it possible to extract exports inside a loader?

### DISCUSSION THREAD

**Discussion by rmwdeveloper** [webpack/webpack#17956] [Q&A]
# Intent
I am trying to use `postcss-loader` with [postcss-js](https://github.com/postcss/postcss-js) as a parser. postcss-js expects a plain javascript object as input, and it returns css. Here is an example lifted directly from their documentation

```javascript
const postcss = require('postcss')
const postcssJs = require('postcss-js')

const style = {
  top: 10,
  '&:hover': {
    top: 5
  }
};

postcss().process(style, { parser: postcssJs }).then( (result) => {
  result.css //=> top: 10px;
             //   &:hover { top: 5px; }
})
```

I would like to author my styles as typescript objects and pass them to postcss + postcss-js.

# Approach
My immediate thought is to use `ts-loader` to compile the typescript down to either CJS or ESM and pass that to postcss-loader to be parsed by postcss-js. 

The only issue though is ts-loader returns the single module compiled by `tsc` as a string, so I need another loader in between `ts-loader` and `postcss-loader` to extract what `postcss-loader` is expecting in this case (a plain javascript object).

In that "in-between" loader, it seems like I pretty much need to parse and execute that string as javascript, follow it's imports, parse and execute those.. etc.. This seems cumbersome and I can't help but think I have the completely wrong approach here.

Here is the relevant portion from my webpack config. This is an excerpt from the module.rules array

```typescript
const config = {
    // ...other stuff
    module: {
        rules: [
            {
                "test": /\.css.ts$/,
                "use": [
                    "style-loader",
                    {
                        "loader": "css-loader",
                        "options": {
                            "importLoaders": 1
                        }
                    },
                    {
                        "loader": "postcss-loader",
                        "options": {
                            "implementation": "postcss",
                            "postcssOptions": {
                                "parser": "postcss-js"
                            }
                        }
                    },
                    "ts-loader"
                ]
            },
            {
                test: /\.tsx?$/,
                use: 'ts-loader',
                exclude: /node_modules/
            }
        ]
    }
};


``` 

And here is the file that is being processed by the webpack rule in question.

Button.css.ts
```typescript
import { Button } from '../styles/Button.styles';

const styles = {
    '.Button': Button
}
export default styles;
```

Button.styles.ts

```typescript
export const Button = {
    display: 'inline-flex'
}
```

# What I've Tried
I can sort of getting it working if I just `require` the code, but hot module reloading doesn't work since the `require` is occurring outside of webpack's context. Here is an example

```typescript
{
  "loader": "postcss-loader",
  "options": {
    "implementation": "postcss",
    "postcssOptions": {
      "parser": function(_, file) {
        const code = require(file.from);
        const postcssjs = require('postcss-js');
        const parsed = postcssjs.parse(code.default);
        return parsed;
      }
    }
  }
}
```

**Comment by alexander-akait** [comment]
I am afraid you will need your own loader here, note - you can result load inside loader
